akka-javasdk/src/main/java/akka/javasdk/OriginAwareContext.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.javasdk;
+
+import akka.javasdk.annotations.Consume;
+import akka.javasdk.eventsourcedentity.EventSourcedEntity;
+import akka.javasdk.keyvalueentity.KeyValueEntity;
+
+import java.util.Optional;
+
+public interface OriginAwareContext extends Context {
+
+
+  /**
+   * When available, this method returns the region where a message was first created.
+   *
+   * <ul>
+   *   <li>It returns a non-empty Optional when consuming events from an {@link EventSourcedEntity}
+   *   or a change updates from a {@link KeyValueEntity}</li>
+   *   <li>It returns an empty Optional when consuming messages from a topic (see {@link Consume.FromTopic})
+   *   or from another service (see {@link Consume.FromServiceStream})</li>
+   * </ul>
+   *
+   * @return the region where a message was first created. When not applicable, it returns an empty Optional.
+   */
+  Optional<String> originRegion();
```

Comment:
- I mentioned this on the runtime side, but maybe I should do it here. In my opinion, it feels more natural to get the originRegion information from the metadata. We can add a helper method extract it, similar to `subject`, `time`, etc., I guess there was a reason to make it this way, what was that?

Reply:
`hasLocalOrigin` is nice, but if that is the only reason, I think users can create such a method themselves if they need it.

Reply:
I think the implementation with meta will also be much simpler because we are passing this information through all the layers from the runtime to the sdk.

Reply:
The reason why I didn't add it to the metadata is because we used to also have JWT and ACL there and we stepped away from it. JWT and ACL is not only present on the endpoints request, which makes more sense.

Also, metadata is available in places where it doesn't make sense to have the `originRegion`, for example in commands. We should have it only when consuming messages.

Reply:
About `hasLocalOrigin`, the method impl is quite simple and indeed anyone can implement this logic, but there is some hidden knowledge in there that otherwise we will need to document.

1. empty means we are consuming from a broker or service-to-service, so not local
   (we will have to document this)
2. for non-empty origins, it needs to match selfRegion, so it depends
   (this is easy and kind of obvious)
3. in dev-mode, both will be "" and therefore local by definition
   (this is also hidden logic that this method is encapsulating)

Reply:
> Also, metadata is available in places where it doesn't make sense to have the originRegion, for example in commands. We should have it only when consuming messages.

Good point, but that implies that we have a leaking abstraction because I can run `metadata.asCloudEvent` in my command handler, which doesn't make any sense. Maybe we should have a separate `CommandMetadata`, `Message/EventMetadata` ?